### DISCUSSION TITLE
What's the best practice to sync workflow status to external DB?

### DISCUSSION THREAD

**Discussion by lpaladin** [argoproj/argo-workflows#11747] [Q&A] [answered]
Hello there!

I'm currently looking for some ideas for syncing "phase" field (WorkflowStatus and NodeStatus) from Argo Workflow object to some external DB for persistence. I would prefer to have the phase status synced at "best effort" quick and achieving eventually consistence.

I'm wondering which approach to go, using workflow hooks to "push" status, or "watch"ing Kubernetes Workflow object. From my view, "push"ing seems unreliable when the call fails, while "watch"ing might cause complexness. Could you please kindly share your thoughts? Thank you in advance!

**Comment by agilgur5** [comment, marked as answer]
There is a built-in feature for NodeStatus that [offloads it to an external DB](https://argoproj.github.io/argo-workflows/offloading-large-workflows/). You might be able to use that for your purposes, though I'm not sure if it would be optimal for or meet your latency requirements. I also don't think it handles WorkflowStatus, just NodeStatus.

With regard to `watch`ing the k8s resources, you could use a [k8s informer](https://www.cncf.io/blog/2019/10/15/extend-kubernetes-via-a-shared-informer/), which is an efficient watch implementation that handles edge-cases and is at the core of most k8s controllers (including Argo itself). Could also use a higher-level tool like [`kubediff`](https://github.com/arriqaaq/kubediff) (or something better maintained, that was just from a quick search).

**EDIT**: You could also set-up a [Resource EventSource](https://argoproj.github.io/argo-events/eventsources/setup/resource/) for this with [Argo Events](https://github.com/argoproj/argo-events).

**Reply by lpaladin** [reply]
Thank you for replying! Seems offloading is only designed for large nodes so that it might not quite fit...

Informer looks promising, but I'm still curious if hooks are also suitable and / or recommended for this use case.
I asked because I'm trying to figure out if Argo team would like users to directly look at K8s resources or setup hooks to do so...

**Reply by agilgur5** [reply]
This is an observation, but I would say the current team is not very opinionated. Argo Workflows itself is also not very opinionated and is very flexible (you might have noticed that there can be several ways to do the same thing, reminiscent of `git` to me).
This is also a pretty specific use-case that I haven't personally seen before (so I don't think a "best practice" exists). Ultimately, as with any software, you're the one working on the specific use-case and know your specific requirements better than anyone else.

You're trying to sync _all_ status changes for _all_ nodes and workflows, so such a hook would need to have a suitable expression and also be one of your [`workflowDefaults`](https://argoproj.github.io/argo-workflows/default-workflow-specs/).
In my personal opinion, that seems more complex and quite tightly coupled. Informers are very generic and serve a variety of purposes, so if your needs evolve, it would not be difficult to adapt that (it's more reusable a piece of software in general). Given your latency requirements, an Informer may also be a better fit, as it does not require running a new Pod, unlike a hook.
_But_ if you don't want to write a separate Go service, if you can create a suitable hook, it might be sufficient for your needs (that's a sizeable _if_ though -- I'm not sure what the expression would look like off the top of my head)

**Reply by agilgur5** [reply]
Oh I forgot, but you could also set-up a [Resource EventSource](https://argoproj.github.io/argo-events/eventsources/setup/resource/) for this with [Argo Events](https://github.com/argoproj/argo-events).

**Reply by lpaladin** [reply]
In this case, Argo Events seems like advanced wrapping around informers and might reduce some of our dev efforts. Thanks for sharing your opinion!
